server/src/main/java/org/opensearch/index/shard/CloseableRetryableRefreshListener.java

```diff
@@ -46,98 +50,163 @@ public CloseableRetryableRefreshListener() {
+     */
+    private void scheduleRetry(TimeValue interval, String retryThreadPoolName, boolean didRefresh) {
+        // If the underlying listener has closed, then we do not allow even the retry to be scheduled
+        if (closed.get() || isRetryEnabled() == false) {
+            return;
+        }
+
+        assert Objects.nonNull(interval) && ThreadPool.THREAD_POOL_TYPES.containsKey(retryThreadPoolName);
+
+        // If the retryScheduled is already true, then we return from here itself. If not, then we proceed with scheduling
+        // the retry.
+        if (retryScheduled.getAndSet(true)) {
             return;
         }
+
         boolean scheduled = false;
         try {
-            this.threadPool.schedule(() -> {
-                boolean successful;
-                boolean permitAcquired = semaphore.tryAcquire();
-                try {
-                    successful = permitAcquired && performAfterRefresh(didRefresh, isRetry);
-                } finally {
-                    if (permitAcquired) {
-                        semaphore.release();
-                    }
-                    retryScheduled.set(false);
-                }
-                scheduleRetry(successful, didRefresh, isRetry || permitAcquired);
-            }, interval, retryThreadPoolName);
+            this.threadPool.schedule(
+                () -> runAfterRefreshWithPermit(didRefresh, () -> retryScheduled.set(false)),
+                interval,
+                retryThreadPoolName
+            );
             scheduled = true;
-            getLogger().info("Scheduled retry with didRefresh={} isRetry={}", didRefresh, isRetry);
+            getLogger().info("Scheduled retry with didRefresh={}", didRefresh);
         } finally {
             if (scheduled == false) {
                 retryScheduled.set(false);
             }
         }
     }
 
+    /**
+     * This returns if the retry is enabled or not. By default, the retries are not enabled.
+     * @return true if retry is enabled.
+     */
+    protected boolean isRetryEnabled() {
+        return false;
+    }
+
+    /**
+     * Runs the performAfterRefresh method under permit. If there are no permits available, then it is no-op. It also hits
+     * the scheduleRetry method with the result value of the performAfterRefresh method invocation.
+     * The synchronised block ensures that if there is a retry or afterRefresh waiting, then it waits until the previous
+     * execution finishes.
+     */
+    private synchronized void runAfterRefreshWithPermit(boolean didRefresh, Runnable runFinally) {
+        if (closed.get()) {
+            return;
+        }
+        boolean successful;
+        boolean permitAcquired = semaphore.tryAcquire();
+        try {
+            successful = permitAcquired && performAfterRefreshWithPermit(didRefresh);
+        } finally {
+            if (permitAcquired) {
+                semaphore.release();
+            }
+            runFinally.run();
```

Comment:
Ideally, this `runFinally` should run before we call `performAfterRefreshWithPermit`(then its name would be different). Having `runAfterRefreshWithPermit` synchronized masks this problem and not an issue with current implementation.

But consider an example where method is not synchronized: if a retry is running and another retry is to be scheduled, the second retry will be skipped.

Reply:
> Ideally, this runFinally should run before we call performAfterRefreshWithPermit(then its name would be different). Having runAfterRefreshWithPermit synchronized masks this problem and not an issue with current implementation.

That's true, for current implementation the afterRefresh or the retry would wait since the `runAfterRefreshWithPermit` is synchronised.